File: app/internal/service/import_service.go
package service

// 导入流程编排：扫描 Basicdata/ → 识别(classify) + Excel财务(excelimport) → 复制归档 → 入库。
// 原子性：文件先写暂存目录，DB 事务 + 磁盘替换全部成功才提交；任一步失败回滚，不破坏现网数据。
// ImportService 依赖注入 *store.Store（事务）与 *config.Config（路径/配置），不再访问包级全局。

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"heritage-mgmt/internal/classify"
	"heritage-mgmt/internal/config"
	"heritage-mgmt/internal/excelimport"
	"heritage-mgmt/internal/store"
)

// ImportStats 导入统计。
type ImportStats struct {
	Units, Projects, Docs, Matched int
}

// ImportService 编排 Basicdata 批量导入（事务 + 扫描目录 + classify + excelimport + 落盘 + 入库）。
type ImportService struct {
	store *store.Store
	cfg   *config.Config
}

// ImportAll 扫描 cfg.AbsBasicdata 导入全部数据；verbose 控制是否打印进度。
// 流程：文件先全部复制到暂存目录（.import_xxx/projects），DB 事务内 ResetTables+入库，
// 全部成功后原子替换 ProjectsDir 并提交事务；任一步失败回滚（现网 ProjectsDir 与 DB 保持不变）。
func (svc *ImportService) ImportAll(verbose bool) (*ImportStats, error) {
	basicdataDir := svc.cfg.AbsBasicdata
	stats := &ImportStats{}
	finRows := excelimport.LoadFinancials(basicdataDir)
	if verbose {
		fmt.Printf("读取明细表, 财务行 %d 条\n", len(finRows))
	}

	entries, err := os.ReadDir(basicdataDir)
	if err != nil {
		return stats, err
	}
	var projDirs []string
	for _, e := range entries {
		if e.IsDir() {
			projDirs = append(projDirs, e.Name())
		}
	}
	sort.Strings(projDirs)

	// 暂存目录：文件先复制到这里，DB+磁盘都成功后再原子替换正式 ProjectsDir。
	stage, err := os.MkdirTemp(svc.cfg.DataDir, ".import_*")
	if err != nil {
		return stats, err
	}
	defer os.RemoveAll(stage) // 成功路径下 projects 已被移走、stage 为空；失败路径下清理半成品
	stageProjects := filepath.Join(stage, "projects")

	unitCache := map[string]int64{}
	unitSort := map[string]int{} // 按导入顺序排序

	tx, err := svc.store.Begin()
	if err != nil {
		return stats, err
	}
	committed := false
	defer func() {
		if !committed {
			_ = tx.Rollback()
		}
	}()
	if err := store.ResetTables(tx); err != nil {
		return stats, err
	}

	for _, folder := range projDirs {
		srcDir := filepath.Join(basicdataDir, folder)
		pname := classify.CleanProjectName(folder)
		unitName, level, category := classify.DetectUnit(pname, svc.cfg.WfCfg.Units.Rules)
		ptype := classify.DetectType(pname, svc.cfg.WfCfg.ProjectTypes.Rules)
		seq := classify.ParseSeq(folder)

		// 单位
		unitID, ok := unitCache[unitName]
		if !ok {
			sortVal := 99
			if s, ok := unitSort[unitName]; ok {
				sortVal = s
			}
			res, err := tx.Exec("INSERT INTO units(name,level,category,sort) VALUES(?,?,?,?)",
				unitName, level, category, sortVal)
			if err != nil {
				return stats, err
			}
			unitID, _ = res.LastInsertId()
			unitCache[unitName] = unitID
			stats.Units++
		}

		fin, score := classify.MatchFinancial(pname, finRows)
		matched := fin != nil

		// 插工程
		res, err := tx.Exec(`INSERT INTO projects(
			unit_id,seq,name,ptype,approval_no,sign_date,complete_date,accept_date,contract_end,
			central_funding,eng_contract,eng_paid,sup_contract,sup_paid,
			des_contract,des_paid,expert_fee,total_paid,status,progress_note,
			source_dir,created) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)`,
			unitID, seq, pname, ptype,
			excelimport.FinGet(fin, "approval_no"),
			excelimport.TrimDate(excelimport.FinGet(fin, "sign_date")), excelimport.TrimDate(excelimport.FinGet(fin, "complete_date")),
			excelimport.TrimDate(excelimport.FinGet(fin, "accept_date")), excelimport.TrimDate(excelimport.FinGet(fin, "contract_end")),
			excelimport.ParseFloat(excelimport.FinGet(fin, "central_funding")),
			excelimport.ParseFloat(excelimport.FinGet(fin, "eng_contract")), excelimport.ParseFloat(excelimport.FinGet(fin, "eng_paid")),
			excelimport.ParseFloat(excelimport.FinGet(fin, "sup_contract")), excelimport.ParseFloat(excelimport.FinGet(fin, "sup_paid")),
			excelimport.ParseFloat(excelimport.FinGet(fin, "des_contract")), excelimport.ParseFloat(excelimport.FinGet(fin, "des_paid")),
			excelimport.ParseFloat(excelimport.FinGet(fin, "expert_fee")), excelimport.ParseFloat(excelimport.FinGet(fin, "total_paid")),
			excelimport.DeriveStatus(fin, svc.cfg.Rules.StatusKeywords), excelimport.FinGet(fin, "progress_note"),
			folder, time.Now().Format("2006-01-02 15:04:05"))
		if err != nil {
			return stats, err
		}
		pid, _ := res.LastInsertId()
		if matched {
			stats.Matched++
		}
		projFolder := fmt.Sprintf("P%04d", pid)
		if _, err := tx.Exec("UPDATE projects SET folder=? WHERE id=?", projFolder, pid); err != nil {
			return stats, err
		}
		destRoot := filepath.Join(stageProjects, projFolder)
		if err := os.MkdirAll(destRoot, 0755); err != nil {
			return stats, err
		}

		// 文件：复制到暂存目录；任一失败立即中断（staging 不提交，现网不受影响）。
		files, err := os.ReadDir(srcDir)
		if err != nil {
			return stats, fmt.Errorf("读取工程目录失败 %s: %w", folder, err)
		}
		for _, fe := range files {
			if fe.IsDir() {
				continue
			}
			fname := fe.Name()
			code, tname := classify.ClassifyDoc(fname, svc.cfg.DocCfg.Types, svc.cfg.DocCfg.UnknownCode, svc.cfg.DocCfg.UnknownName)
			destDir := filepath.Join(destRoot, code)
			if err := os.MkdirAll(destDir, 0755); err != nil {
				return stats, err
			}
			dst := filepath.Join(destDir, fname)
			if err := copyFile(filepath.Join(srcDir, fname), dst); err != nil {
				return stats, fmt.Errorf("复制文件失败 %s: %w", fname, err)
			}
			rel, _ := filepath.Rel(stageProjects, dst)
			rel = filepath.ToSlash(rel)
			ext := strings.ToLower(strings.TrimPrefix(filepath.Ext(fname), "."))
			fi, _ := os.Stat(dst)
			var size int64
			if fi != nil {
				size = fi.Size()
			}
			_, err := tx.Exec(`INSERT INTO documents(
				project_id,doc_type,doc_type_name,title,orig_name,
				file_path,file_ext,file_size,uploaded,source) VALUES(?,?,?,?,?,?,?,?,?,?)`,
				pid, code, tname, classify.CleanTitle(fname), fname, rel, ext, size,
				time.Now().Format("2006-01-02 15:04:05"), "import")
			if err != nil {
				return stats, err
			}
			stats.Docs++
		}
		stats.Projects++
		if verbose {
			tag := ""
			if matched {
				amt := excelimport.FinGet(fin, "central_funding")
				tag = fmt.Sprintf(" [财务 score=%.2f 指标=%s万]", score, amt)
			} else {
				tag = " [未匹配财务]"
			}
			fmt.Printf("  OK %s | %s | %s%s\n", unitName, pname, ptype, tag)
		}
	}

	// 原子替换 ProjectsDir：旧目录改名备份 → 暂存 projects 就位 → 提交 DB；任一步失败回滚。
	oldProjects := svc.cfg.ProjectsDir + "_pre_import_" + time.Now().Format("20060102_150405")
	if err := os.Rename(svc.cfg.ProjectsDir, oldProjects); err != nil && !os.IsNotExist(err) {
		return stats, err
	}
	if err := os.Rename(stageProjects, svc.cfg.ProjectsDir); err != nil {
		_ = os.Rename(oldProjects, svc.cfg.ProjectsDir) // 回滚：恢复旧目录
		return stats, err
	}
	if err := tx.Commit(); err != nil {
		// DB 提交失败：移走新目录、恢复旧目录
		_ = os.RemoveAll(svc.cfg.ProjectsDir)
		_ = os.Rename(oldProjects, svc.cfg.ProjectsDir)
		return stats, err
	}
	committed = true
	_ = os.RemoveAll(oldProjects) // 成功：清理旧目录

	if verbose {
		fmt.Printf("\n导入完成: 单位 %d / 工程 %d / 文档 %d / 财务匹配 %d\n",
			stats.Units, stats.Projects, stats.Docs, stats.Matched)
	}
	return stats, nil
}

// copyFile 复制单个文件，检查 Close 错误以保证落盘完整。
func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()
	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	if _, err := io.Copy(out, in); err != nil {
		out.Close()
		return err
	}
	return out.Close()
}

File: desktop/src/core/import/ImportService.cpp
#include "ImportService.h"

#include "core/classify/Classify.h"
#include "core/excelimport/ExcelImport.h"
#include "core/storage/LogRepo.h"
#include "core/storage/Schema.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QSqlError>
#include <QSqlQuery>
#include <QVariant>
#include <algorithm>
#include <optional>

namespace heritage {

namespace {
void bindOpt(QSqlQuery& q, const std::optional<double>& v) {
    q.addBindValue(v.has_value() ? QVariant(*v) : QVariant());
}
} // namespace

ImportService::ImportService(QSqlDatabase db, const AppConfig& cfg, LogRepo& logs)
    : db_(db), cfg_(cfg), logs_(logs) {}

ImportStats ImportService::importAll(const QString& basicdataDir) {
    ImportStats st;
    err_.clear();

    const QVector<QMap<QString, QString>> finRows = excel::loadFinancials(basicdataDir);

    // 子目录（每个 = 一个工程文件夹）
    QDir bd(basicdataDir);
    QStringList projDirs;
    for (const QFileInfo& e : bd.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot))
        projDirs << e.fileName();
    std::sort(projDirs.begin(), projDirs.end());

    if (!db_.transaction()) {
        err_ = QStringLiteral("开启事务失败");
        return st;
    }
    if (!schema::resetTables(db_)) {
        err_ = QStringLiteral("清表失败");
        db_.rollback();
        return st;
    }

    QHash<QString, qint64> unitCache;
    const QString now = QDateTime::currentDateTime().toString(QStringLiteral("yyyy-MM-dd HH:mm:ss"));
    QSqlQuery q(db_);

    for (const QString& folder : projDirs) {
        const QString srcDir = basicdataDir + QStringLiteral("/") + folder;
        const QString pname = classify::cleanProjectName(folder);
        const auto ud = classify::detectUnit(pname, cfg_.workflow.unitRules);
        const QString ptype = classify::detectType(pname, cfg_.workflow.typeRules);
        const auto seq = classify::parseSeq(folder);

        // 单位（缓存，sort 固定 99 —— Go unitSort 实为死代码）
        qint64 unitId = 0;
        const auto cit = unitCache.constFind(ud.unit);
        if (cit != unitCache.cend()) {
            unitId = cit.value();
        } else {
            q.prepare(QStringLiteral("INSERT INTO units(name,level,category,sort) VALUES(?,?,?,?)"));
            q.addBindValue(ud.unit);
            q.addBindValue(ud.level);
            q.addBindValue(ud.category);
            q.addBindValue(99);
            if (!q.exec()) {
                err_ = QStringLiteral("插单位失败: ") + q.lastError().text();
                db_.rollback();
                return st;
            }
            unitId = q.lastInsertId().toLongLong();
            unitCache.insert(ud.unit, unitId);
            st.units++;
        }

        // 财务匹配
        const auto fm = classify::matchFinancial(pname, finRows);
        const bool matched = fm.index >= 0;
        const QMap<QString, QString> fin = matched ? finRows.at(fm.index) : QMap<QString, QString>();

        // 插工程（22 字段；accept_date 复用 complete_date，与 Go 一致）
        q.prepare(QStringLiteral(
            "INSERT INTO projects(unit_id,seq,name,ptype,approval_no,sign_date,complete_date,"
            "accept_date,contract_end,central_funding,eng_contract,eng_paid,sup_contract,sup_paid,"
            "des_contract,des_paid,expert_fee,total_paid,status,progress_note,source_dir,created) "
            "VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"));
        q.addBindValue(unitId);
        q.addBindValue(seq.has_value() ? QVariant(*seq) : QVariant());
        q.addBindValue(pname);
        q.addBindValue(ptype);
        q.addBindValue(excel::finGet(fin, QStringLiteral("approval_no")));
        q.addBindValue(excel::trimDate(excel::finGet(fin, QStringLiteral("sign_date"))));
        q.addBindValue(excel::trimDate(excel::finGet(fin, QStringLiteral("complete_date"))));
        q.addBindValue(excel::trimDate(excel::finGet(fin, QStringLiteral("complete_date"))));
        q.addBindValue(excel::trimDate(excel::finGet(fin, QStringLiteral("contract_end"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("central_funding"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("eng_contract"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("eng_paid"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("sup_contract"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("sup_paid"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("des_contract"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("des_paid"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("expert_fee"))));
        bindOpt(q, excel::parseFloat(excel::finGet(fin, QStringLiteral("total_paid"))));
        q.addBindValue(excel::deriveStatus(fin, cfg_.rules.statusKeywords));
        q.addBindValue(excel::finGet(fin, QStringLiteral("progress_note")));
        q.addBindValue(folder);
        q.addBindValue(now);
        if (!q.exec()) {
            err_ = QStringLiteral("插工程失败: ") + q.lastError().text();
            db_.rollback();
            return st;
        }
        const qint64 pid = q.lastInsertId().toLongLong();
        if (matched)
            st.matched++;
        const QString projFolder = QStringLiteral("P%1").arg(pid, 4, 10, QLatin1Char('0'));
        q.prepare(QStringLiteral("UPDATE projects SET folder=? WHERE id=?"));
        q.addBindValue(projFolder);
        q.addBindValue(pid);
        q.exec();
        const QString destRoot = cfg_.projectsDir + QStringLiteral("/") + projFolder;
        QDir().mkpath(destRoot);

        // 归档文件
        QDir src(srcDir);
        for (const QFileInfo& fe : src.entryInfoList(QDir::Files)) {
            const QString fname = fe.fileName();
            const auto dc = classify::classifyDoc(fname, cfg_.docCfg.types,
                                                  cfg_.docCfg.unknownCode, cfg_.docCfg.unknownName);
            const QString destDir = destRoot + QStringLiteral("/") + dc.code;
            QDir().mkpath(destDir);
            const QString dst = destDir + QStringLiteral("/") + fname;
            if (!QFile::copy(srcDir + QStringLiteral("/") + fname, dst))
                continue; // 复制失败告警跳过
            QString rel = QDir(cfg_.projectsDir).relativeFilePath(dst);
            rel.replace(QLatin1Char('\\'), QLatin1Char('/'));
            const QString ext = fe.suffix().toLower();
            const qint64 size = QFileInfo(dst).size();
            q.prepare(QStringLiteral(
                "INSERT INTO documents(project_id,doc_type,doc_type_name,title,orig_name,"
                "file_path,file_ext,file_size,uploaded,source) VALUES(?,?,?,?,?,?,?,?,?,?)"));
            q.addBindValue(pid);
            q.addBindValue(dc.code);
            q.addBindValue(dc.name);
            q.addBindValue(classify::cleanTitle(fname));
            q.addBindValue(fname);
            q.addBindValue(rel);
            q.addBindValue(ext);
            q.addBindValue(size);
            q.addBindValue(now);
            q.addBindValue(QStringLiteral("import"));
            if (!q.exec()) {
                err_ = QStringLiteral("插文档失败: ") + q.lastError().text();
                db_.rollback();
                return st;
            }
            st.docs++;
        }
        st.projects++;
    }

    if (!db_.commit()) {
        err_ = QStringLiteral("提交事务失败");
        db_.rollback();
        return st;
    }
    logs_.insert(QStringLiteral("批量导入"), basicdataDir,
                 QStringLiteral("单位%1/工程%2/文档%3").arg(st.units).arg(st.projects).arg(st.docs));
    return st;
}

} // namespace heritage
